### ISSUE TITLE
cmake triplet selection (VCPKG_TARGET_TRIPLET) not behave as wanted

### ISSUE THREAD

**Issue by nerditation** [Microsoft/vcpkg#5149]
I wanted to try out vcpkg. I created a simple project with two files: ``main.cpp`` and ``CMakeLists.txt``

main.cpp:
```cpp
#include <lua.hpp>
int main (int, char **) {
    auto L = luaL_newstate();
    luaL_openlibs(L);
    luaL_dostring(L, "print 'hello world'");
    lua_close(L)
    return 0;
}
```

CMakeLists.txt:
```cmake
project(lunatic)
add_executable(lunatic main.cpp)
find_package(lua 5.3 REQUIRED)
include_directories(${LUA_INCULDE_DIR})
target_link_libraries(lunatic ${LUA_LIBRARIES})
```

really straight forward, I used this to test my build environment on a Linux system.
so I didn't expect anything would go wrong on Windows. but when I build, it fails:
<details>
<summary>
<p>click to see the full log:</p>
<pre>
Z:\lunatic\main.cpp(2): fatal error C1083: Cannot open include file: 'lua.hpp': No such file or directory </pre>
</summary>
<pre>
Z:\build>cmake -DCMAKE_TOOLCHAIN_FILE=W:\vcpkg\scripts\buildsystems\vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x86-windows-static ..\lunatic
-- Building for: Visual Studio 15 2017
-- The C compiler identification is MSVC 19.16.27026.1
-- The CXX compiler identification is MSVC 19.16.27026.1
-- Check for working C compiler: D:/Applications/Microsoft/VS2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x86/cl.exe
-- Check for working C compiler: D:/Applications/Microsoft/VS2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x86/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: D:/Applications/Microsoft/VS2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x86/cl.exe
-- Check for working CXX compiler: D:/Applications/Microsoft/VS2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x86/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Lua: W:/vcpkg/installed/x86-windows-static/debug/lib/lua.lib (found version "5.3.4")
CMake Warning (dev) in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 3.13)

  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done
-- Generating done
-- Build files have been written to: Z:/build

Z:\build>cmake --build .
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 2019/1/14 22:44:01.
Project "Z:\build\ALL_BUILD.vcxproj" on node 1 (default targets).
Project "Z:\build\ALL_BUILD.vcxproj" (1) is building "Z:\build\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
PrepareForBuild:
  Creating directory "Win32\Debug\ZERO_CHECK\".
  Creating directory "Z:\build\Debug\".
  Creating directory "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\".
InitializeBuildStatus:
  Creating "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Checking Build System
  CMake does not need to re-run because Z:/build/CMakeFiles/generate.stamp is up-to-date.
FinalizeBuildStatus:
  Deleting file "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "Z:\build\ZERO_CHECK.vcxproj" (default targets).

Project "Z:\build\ALL_BUILD.vcxproj" (1) is building "Z:\build\lunatic.vcxproj" (3) on node 1 (default targets).
PrepareForBuild:
  Creating directory "lunatic.dir\Debug\".
  Creating directory "lunatic.dir\Debug\lunatic.tlog\".
InitializeBuildStatus:
  Creating "lunatic.dir\Debug\lunatic.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule Z:/lunatic/CMakeLists.txt
  CMake does not need to re-run because Z:/build/CMakeFiles/generate.stamp is up-to-date.
VcpkgTripletSelection:
  Not using Vcpkg because VcpkgEnabled is "false"
ClCompile:
  D:\Applications\Microsoft\VS2017\BuildTools\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\CL.exe /c /Zi /nologo /W3 /WX- /diagnostics:classic /Od /Ob0 /Oy- /D WIN32 /D _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"lunatic.dir\Debug\\" /Fd"lunatic.dir\Debug\vc141.pdb" /Gd /TP /analyze- /errorReport:queue Z:\lunatic\main.cpp
  main.cpp
Z:\lunatic\main.cpp(2): fatal error C1083: Cannot open include file: 'lua.hpp': No such file or directory [Z:\build\lunatic.vcxproj]
Done Building Project "Z:\build\lunatic.vcxproj" (default targets) -- FAILED.

Done Building Project "Z:\build\ALL_BUILD.vcxproj" (default targets) -- FAILED.

Build FAILED.

"Z:\build\ALL_BUILD.vcxproj" (default target) (1) ->
"Z:\build\lunatic.vcxproj" (default target) (3) ->
(ClCompile target) ->
  Z:\lunatic\main.cpp(2): fatal error C1083: Cannot open include file: 'lua.hpp': No such file or directory [Z:\build\lunatic.vcxproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:09.48
</pre>
</details>

I noticed this line:
```
VcpkgTripletSelection:
  Not using Vcpkg because VcpkgEnabled is "false"
```

after some searching, I find #5064, which is not merged yet. ok, fine, I don't know how it works, but it's just a two-line patch, not a big deal.

now try again, it built! wait, I noticed something weird:

<details>
<summary>
<p>click to see the full log:</p>
<pre>
VcpkgTripletSelection:
  Using triplet "x86-windows" from "W:\vcpkg\installed\x86-windows\"
</pre>
</summary>
<pre>
Z:\build>cmake --build .
CMake is re-running because Z:/build/CMakeFiles/generate.stamp is out-of-date.
  the file 'W:/vcpkg/scripts/buildsystems/vcpkg.cmake'
  is newer than 'Z:/build/CMakeFiles/generate.stamp.depend'
  result='-1'
CMake Warning (dev) in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 3.13)

  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done
-- Generating done
-- Build files have been written to: Z:/build
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 2019/1/14 23:04:15.
Project "Z:\build\ALL_BUILD.vcxproj" on node 1 (default targets).
Project "Z:\build\ALL_BUILD.vcxproj" (1) is building "Z:\build\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
InitializeBuildStatus:
  Creating "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  All outputs are up-to-date.
FinalizeBuildStatus:
  Deleting file "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "Z:\build\ZERO_CHECK.vcxproj" (default targets).

Project "Z:\build\ALL_BUILD.vcxproj" (1) is building "Z:\build\lunatic.vcxproj" (3) on node 1 (default targets).
InitializeBuildStatus:
  Touching "lunatic.dir\Debug\lunatic.tlog\unsuccessfulbuild".
CustomBuild:
  All outputs are up-to-date.
VcpkgTripletSelection:
  Using triplet "x86-windows" from "W:\vcpkg\installed\x86-windows\"
ClCompile:
  D:\Applications\Microsoft\VS2017\BuildTools\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\CL.exe /c /I"W:\vcpkg\installed\x86-windows\include" /Zi /nologo /W3 /WX- /diagnostics:classic /Od /Ob0 /Oy- /D WIN32 /D _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t
  /Zc:forScope /Zc:inline /GR /Fo"lunatic.dir\Debug\\" /Fd"lunatic.dir\Debug\vc141.pdb" /Gd /TP /analyze- /errorReport:queue Z:\lunatic\main.cpp
  main.cpp
Link:
  D:\Applications\Microsoft\VS2017\BuildTools\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\link.exe /ERRORREPORT:QUEUE /OUT:"Z:\build\Debug\lunatic.exe" /INCREMENTAL /NOLOGO /LIBPATH:"W:\vcpkg\installed\x86-windows\debug\lib" /LIBPATH:"W:\vcpkg\installed\x86-windows\debug\lib\manual-link" "W:\vcpkg\installed\x86-windows-static\debug\lib\lua.lib" kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib "W:\vcpkg\installed\x86-windows\debug\lib\*.lib" /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"Z:/build/Debug/lunatic.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"Z:/build/Debug/lunatic.lib" /MACHINE:X86 /SAFESEH  /machine:X86 lunatic.dir\Debug\main.obj
  lunatic.vcxproj -> Z:\build\Debug\lunatic.exe
AppLocalFromInstalled:
  C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy Bypass -noprofile -File "W:\vcpkg\scripts\buildsystems\msbuild\applocal.ps1" "Z:\build\Debug\lunatic.exe" "W:\vcpkg\installed\x86-windows\debug\bin" "lunatic.dir\Debug\lunatic.tlog\lunatic.write.1u.tlog" "lunatic.dir\Debug\vcpkg.applocal.log"
  Z:\build\Debug\lua.dll
_CopyFilesMarkedCopyLocal:
  Creating "Z:\build\lunatic.dir\Debug\lunatic.vcxproj.CopyComplete" because "AlwaysCreate" was specified.
PostBuildEvent:
  setlocal
  powershell -noprofile -executionpolicy Bypass -file W:/vcpkg/scripts/buildsystems/msbuild/applocal.ps1 -targetBinary Z:/build/Debug/lunatic.exe -installedDir W:/vcpkg/installed/x86-windows-static/debug/bin -OutVariable out
  if %errorlevel% neq 0 goto :cmEnd
  :cmEnd
  endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone
  :cmErrorLevel
  exit /b %1
  :cmDone
  if %errorlevel% neq 0 goto :VCEnd
  :VCEnd
FinalizeBuildStatus:
  Deleting file "lunatic.dir\Debug\lunatic.tlog\unsuccessfulbuild".
  Touching "lunatic.dir\Debug\lunatic.tlog\lunatic.lastbuildstate".
Done Building Project "Z:\build\lunatic.vcxproj" (default targets).

PrepareForBuild:
  Creating directory "Win32\Debug\ALL_BUILD\".
  Creating directory "Win32\Debug\ALL_BUILD\ALL_BUILD.tlog\".
InitializeBuildStatus:
  Creating "Win32\Debug\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule Z:/lunatic/CMakeLists.txt
  CMake does not need to re-run because Z:/build/CMakeFiles/generate.stamp is up-to-date.
FinalizeBuildStatus:
  Deleting file "Win32\Debug\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild".
  Touching "Win32\Debug\ALL_BUILD\ALL_BUILD.tlog\ALL_BUILD.lastbuildstate".
Done Building Project "Z:\build\ALL_BUILD.vcxproj" (default targets).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:13.02
</pre>
</details>

I thought to select a triplet is as simple as a command line argument, like the example in the [document/blogpost](https://blogs.msdn.microsoft.com/vcblog/2016/11/01/vcpkg-updates-static-linking-is-now-available/), but apparently the triplet is different from what I specified via command line. 
but just to make sure, let me double check the variable is indeed set:
<pre>
Z:\build>grep VCPKG_TARGET_TRIPLET CMakeCache.txt
VCPKG_TARGET_TRIPLET:STRING=x86-windows-static
</pre>

if we look more closely at the linker command:
<pre>
 D:\Applications\Microsoft\VS2017\BuildTools\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\link.exe /ERRORREPORT:QUEUE /OUT:"Z:\build\Debug\lunatic.exe" /INCREMENTAL /NOLOGO /LIBPATH:"W:\vcpkg\installed\x86-windows\debug\lib" /LIBPATH:"W:\vcpkg\installed\x86-windows\debug\lib\manual-link" "W:\vcpkg\installed\x86-windows-static\debug\lib\lua.lib" kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib "W:\vcpkg\installed\x86-windows\debug\lib\*.lib" /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"Z:/build/Debug/lunatic.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"Z:/build/Debug/lunatic.lib" /MACHINE:X86 /SAFESEH  /machine:X86 lunatic.dir\Debug\main.obj
</pre>

the LIBPATH is from "x86-windows", but there is a explicit lua.lib file specified from the "x86-windows-static".
my guess was that the explicit lua.lib is added by the ``target_link_library`` cmake command
while the LIBPATH and wildcard *.lib are selected by msbuild.

I don't know anything about how msbuild and vcpkg work, but I managed to figure out a solution.
first I searched ``VcpkgTripletSelection`` and found it located in the ``vcpkg.targets`` file:
```xml
  <Target Name="VcpkgTripletSelection" BeforeTargets="ClCompile">
    <Message Text="Using triplet &quot;$(VcpkgTriplet)&quot; from &quot;$(VcpkgRoot)&quot;" Importance="Normal" Condition="'$(VcpkgEnabled)' == 'true'"/>
    <Message Text="Not using Vcpkg because VcpkgEnabled is &quot;$(VcpkgEnabled)&quot;" Importance="Normal" Condition="'$(VcpkgEnabled)' != 'true'"/>
    <Message Text="Vcpkg is unable to link because we cannot decide between Release and Debug libraries. Please define the property VcpkgConfiguration to be 'Release' or 'Debug' (currently '$(VcpkgConfiguration)')." Importance="High" Condition="'$(VcpkgEnabled)' == 'true' and '$(VcpkgNormalizedConfiguration)' == ''"/>
  </Target>
```
after reading this file, to my understanding, it basically check two properties ``VcpkgEnabled`` and ``VcpkgTriplet``, and guesses some default value in case they are absent.

that ``VcpkgEnabled`` looks familiar... isn't it the one mentioned in #5064? I don't know cmake very well, but I can guess the VS_GLOBAL_xxx property is the magic word. so I ended up with these modifications to ``vcpkg.cmake``.
```diff
diff --git a/scripts/buildsystems/vcpkg.cmake b/scripts/buildsystems/vcpkg.cmake
index 91c196fb..4073a66f 100644
--- a/scripts/buildsystems/vcpkg.cmake
+++ b/scripts/buildsystems/vcpkg.cmake
@@ -151,7 +151,8 @@ function(add_executable name)
             )
         endif()
         set_target_properties(${name} PROPERTIES VS_USER_PROPS do_not_import_user.props)
-        set_target_properties(${name} PROPERTIES VS_GLOBAL_VcpkgEnabled false)
+        set_target_properties(${name} PROPERTIES VS_GLOBAL_VcpkgEnabled true)
+        set_target_properties(${name} PROPERTIES VS_GLOBAL_VcpkgTriplet ${VCPKG_TARGET_TRIPLET})
     endif()
 endfunction()

@@ -171,7 +172,8 @@ function(add_library name)
             )
         endif()
         set_target_properties(${name} PROPERTIES VS_USER_PROPS do_not_import_user.props)
-        set_target_properties(${name} PROPERTIES VS_GLOBAL_VcpkgEnabled false)
+        set_target_properties(${name} PROPERTIES VS_GLOBAL_VcpkgEnabled true)
+        set_target_properties(${name} PROPERTIES VS_GLOBAL_VcpkgTriplet ${VCPKG_TARGET_TRIPLET})
     endif()
 endfunction()
```

after the changes were made, it finally worked (with a ``LNK4098`` warning regarding dynamic vs static crt linking, but this, at least, I understand it and know the solution to it)

I don't know if this is indeed a bug in vcpkg, or I am just missing something really obvious.
but it is frustrating that even such simple code can't "just work" out of the box.

**Comment by nerditation** [comment]
ah silly me. the first time it didn't find the include file is because I made a typo in the ``CMakeLists.txt``. but the triplet selection part of this post is still relevant. after I fixed the typo and cleaned the cache, the build still select ``x86-windows``, as opposed to ``x86-windows-static``.

it's funny that if I didn't make the typo in the first place, I probably wouldn't find the #5064, which leads me to find the cmake special ``VS_GLOBAL_xxx`` **magic**.

**Comment by nerditation** [comment]
after several hours of playing around with vcpkg, I think I finally got a rough idea of how vcpkg interact with cmake and msbuild.

if I'm understanding correctly, the ``vcpkg.targets`` file is only evaluated by msbuild when vcpkg integration is installed. at the same time, the cmake toolchain file ``vcpkg.cmake`` is doing the same thing for cmake based project.

if a project uses the cmake toolchain file, and configured to generate the msbuild building back end, depending on the actual configuration, the two systems are redundant at best (as the final solution in my original post), or in conflict at worst (as my first attempt after applying #5064).


I think this is why the property "VcpkgEnabled" was introduced in the first place. it is there to prevent cmake generated msbuild project to evaluate the vcpkg msbuild integration again. so neither the change in my original post are nessesary, nor the #5064 would get merged at all.

I tried to edit my original post the delete the first part about the typo, then I think it'd be better to leave the original post as is to give some context of my mistakes for later readers of this thread.

now that I figured it out, I'll just close this thread for now. sorry for the noise.
